pkg/services/ngalert/models/testing.go

```diff
@@ -20,320 +19,477 @@ import (
 		}
 	}
 }
 
-func WithQuery(query ...AlertQuery) AlertRuleMutator {
-	return func(rule *AlertRule) {
+func (a *AlertRuleMutators) WithQuery(query ...AlertQuery) AlertRuleMutator {
+	return func(_ *AlertRuleGenerator, rule *AlertRule) {
 		rule.Data = query
 		if len(query) > 1 {
 			rule.Condition = query[0].RefID
 		}
 	}
 }
 
-func WithGroupKey(groupKey AlertRuleGroupKey) AlertRuleMutator {
-	return func(rule *AlertRule) {
+func (a *AlertRuleMutators) WithGroupName(groupName string) AlertRuleMutator {
+	return func(_ *AlertRuleGenerator, rule *AlertRule) {
+		rule.RuleGroup = groupName
+	}
+}
+
+func (a *AlertRuleMutators) WithGroupPrefix(prefix string) AlertRuleMutator {
+	return func(g *AlertRuleGenerator, rule *AlertRule) {
+		rule.RuleGroup = fmt.Sprintf("%s%s", prefix, util.GenerateShortUID())
+	}
+}
+
+func (a *AlertRuleMutators) WithGroupKey(groupKey AlertRuleGroupKey) AlertRuleMutator {
+	return func(_ *AlertRuleGenerator, rule *AlertRule) {
 		rule.RuleGroup = groupKey.RuleGroup
 		rule.OrgID = groupKey.OrgID
 		rule.NamespaceUID = groupKey.NamespaceUID
 	}
 }
 
-func WithNotificationSettingsGen(ns func() NotificationSettings) AlertRuleMutator {
-	return func(rule *AlertRule) {
+func (a *AlertRuleMutators) WithSameGroup() AlertRuleMutator {
+	once := sync.Once{}
+	name := ""
+	return func(g *AlertRuleGenerator, rule *AlertRule) {
+		once.Do(func() {
+			name = util.GenerateShortUID()
+		})
+		rule.RuleGroup = name
+	}
+}
+
+func (a *AlertRuleMutators) WithNotificationSettingsGen(ns func() NotificationSettings) AlertRuleMutator {
+	return func(_ *AlertRuleGenerator, rule *AlertRule) {
 		rule.NotificationSettings = []NotificationSettings{ns()}
 	}
 }
+func (a *AlertRuleMutators) WithNotificationSettings(ns NotificationSettings) AlertRuleMutator {
+	return func(_ *AlertRuleGenerator, rule *AlertRule) {
+		rule.NotificationSettings = []NotificationSettings{ns}
+	}
+}
 
-func WithNoNotificationSettings() AlertRuleMutator {
-	return func(rule *AlertRule) {
+func (a *AlertRuleMutators) WithNoNotificationSettings() AlertRuleMutator {
+	return func(_ *AlertRuleGenerator, rule *AlertRule) {
 		rule.NotificationSettings = nil
 	}
 }
 
-func GenerateAlertLabels(count int, prefix string) data.Labels {
+func (a *AlertRuleMutators) WithIsPaused(paused bool) AlertRuleMutator {
+	return func(_ *AlertRuleGenerator, rule *AlertRule) {
+		rule.IsPaused = paused
+	}
+}
+
+func (g *AlertRuleGenerator) GenerateLabels(min, max int, prefix string) data.Labels {
+	count := max
+	if min > max {
+		panic("min should not be greater than max")
+	} else if min < max {
+		count = g.Rand.Intn(max-min) + min
+	}
 	labels := make(data.Labels, count)
 	for i := 0; i < count; i++ {
 		labels[prefix+"key-"+util.GenerateShortUID()] = prefix + "value-" + util.GenerateShortUID()
 	}
 	return labels
 }
 
+func GenerateAlertLabels(count int, prefix string) data.Labels {
+	return gen.GenerateLabels(count, count, prefix)
```

Comment:
If you were to make the global rand change I suggested, the bodies of this and `g.GenerateLabels` could basically be swapped.